registry/storage/metadata/transaction.go

```diff
@@ -0,0 +1,164 @@
+package metadata
+
+import (
+	"fmt"
+
+	"errors"
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/uuid"
+)
+
+// TxFunc runs a set of operations within a transaction
+type MetadataUpdateFunc func(context.Context) error
+
+var ErrTransactionCanRetry = errors.New("transaction can be retried")
+
+// Update takes a function which makes a series of mutations and tries
+// to commit it
+func Update(ctx context.Context, repo distribution.Repository, f MetadataUpdateFunc) error {
+	metadataRepo, support := repo.(Metadatable)
+	if !support {
+		// no metadata store configured
+		return f(ctx)
+	}
+
+	tx, err := BeginTx(metadataRepo.MetadataService())
+	if err != nil {
+		return err
+	}
+	ctx = WithTx(ctx, tx)
+
+	var attempts int
+try:
+	// Prepare the transaction
+	err = f(ctx)
+	if err != nil {
+		return err
+	}
+
+	var txErr error
+	if attempts == 1 {
+		return fmt.Errorf("transaction failed: %s", txErr)
+	}
+
+	if txErr = tx.Commit(ctx); txErr != nil {
+		if canRetry(txErr) {
+			attempts++
+			goto try
+		}
+	}
+
+	return txErr
+}
+
+func canRetry(err error) bool {
+	return err == ErrTransactionCanRetry
+}
+
+type Transaction struct {
+	id        uuid.UUID
+	store     MetadataService
+	prepared  map[string]MetadataUpdateRecord
+	committed bool
+}
+
+// todo: rollback state
+
+func BeginTx(store MetadataService) (*Transaction, error) {
+	return &Transaction{
+		id:       uuid.Generate(),
+		prepared: make(map[string]MetadataUpdateRecord),
+		store:    store,
+	}, nil
+}
+
+func (t *Transaction) Update(ctx context.Context, key string, val interface{}) error {
+	if t.committed {
+		return fmt.Errorf("Unable to update %s, already committed", t.id)
+	}
+
+	expected, err := t.store.Get(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	u := MetadataUpdateRecord{Actual: val, Expected: expected}
+	t.prepared[key] = u
+
+	valstr := fmt.Sprint(val)
+	if len(valstr) > 20 {
+		valstr = valstr[:20] + "..."
+	}
+	curstr := fmt.Sprint(expected)
+	if len(curstr) > 20 {
+		curstr = curstr[:20] + "..."
+	}
+	fmt.Printf("Updated transaction %s with %q=%s, cur=%s\n", t.id, key, valstr, curstr)
+	return nil
+}
+
+func (t *Transaction) Commit(ctx context.Context) error {
+	if t.committed {
+		return fmt.Errorf("Unable to commit %s, already committed", t.id)
+	}
+
+	err := t.store.BatchPut(ctx, t.prepared)
+	if err != nil {
+		return err
+	}
```

Comment:
I'm wondering if building the transactionality around `BatchPut` is the right design.

If two transactions step on each other, one will fail and need to retry. But it may never succeed if it keeps getting stepped on. If there's a limit on retries, eventually the operation will fail.

It feels like the behavior would be more deterministic if a write transaction took a lock over the relevant data until `Commit` time. This could also dramatically reduce latency because you wouldn't have to do a `Get` for every `Update` (and repeat those gets during `BatchPut` to validate that the transaction hasn't been stepped on).

Our application is quite different so not all of the design decisions will make sense here, but it may be interesting to look at the raft store from SwarmKit:

https://github.com/docker/swarmkit/blob/master/manager/state/store/memory.go

It allows one write transaction at a time. It also supports an unlimited number of read transactions, which work from a consistent snapshot of the data set (since our local copy of the data is a radix tree, snapshotting is a basically a pointer copy).

We version our objects for convenience. This doesn't matter when only a single write transaction is involved, because consistency is guaranteed in that case. But if you use a read transaction to retrieve an object, then later try to update that object inside a write transaction and it has changed since you read it, the update will fail because the version number is not up to date. Versioning might be helpful here if you stay on the path of doing transactions based on conflict detection, because in case the objects you're storing are large, you wouldn't have to retrieve and deserialize them to check for a conflict.